### PR TITLE
API to get single commit via SHA and Ref

### LINKED ISSUE #10089: [Feature] [API] Get a single commit
doc ref: https://developer.github.com/v3/repos/commits/#get-a-single-commit

https://github.com/go-gitea/gitea/blob/32fb813133c74ddc3af1964e81fff72fea4f24f1/modules/convert/convert.go#L318

### PR DESCRIPTION
**Pull request by 6543**
close #10089

`GET /repos/:owner/:repo/commits/:ref`

* [x] implement func
* [x] check ref/sha if it is one
* [x] write Tests

### CODE REVIEW

**Comment by 6543** [comment]
## GH reference: https://developer.github.com/v3/repos/commits/#get-a-single-commit

File: routers/api/v1/repo/commits.go [hunk 1]
```diff
@@ -69,6 +69,64 @@ func GetSingleCommit(ctx *context.APIContext) {
 	ctx.JSON(http.StatusOK, json)
 }
 
+// GetSingleCommitByRef get a commit via ref
+func GetSingleCommitByRef(ctx *context.APIContext) {
```
Comment (lunny) [inline routers/api/v1/repo/commits.go:64]: It seems the two function have most same codes?

Comment (6543) [reply routers/api/v1/repo/commits.go:64]: they do

**Comment by 6543** [comment]
@lunny you could use `GetSingleCommitBySHA` for Ref too since the internal function doesnt care ...
but to make you API more similar to GitHub's one ... I created the api point

File: routers/api/v1/repo/commits.go [hunk 2]
```diff
@@ -48,13 +48,61 @@ func GetSingleCommit(ctx *context.APIContext) {
 	//   "404":
 	//     "$ref": "#/responses/notFound"
 
+	sha := ctx.Params(":sha")
+	if len(sha) == 0 {
```
Comment (guillep2k) [inline routers/api/v1/repo/commits.go]: I'm sure there's a function that checks if a string is a valid commit sha (i.e. accept full and shortened ones). We should perform that validation here.

File: routers/api/v1/repo/commits.go [hunk 3]
```diff
@@ -48,13 +48,61 @@ func GetSingleCommit(ctx *context.APIContext) {
...
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/Commit"
+	//   "400":
+	//     "$ref": "#/responses/error"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+
+	ref := ctx.Params("ref")
+	if len(ref) == 0 {
```
Comment (guillep2k) [inline routers/api/v1/repo/commits.go]: Maybe we should validate `ref` value as well? I did some tests and so far the only invalid strings I could generate were those:

- starting or ending with `/`
- containing spaces, `:`, `*`, `[`, `~`, `\` or `?`.
- containing `"` (it's accepted, but the command fails).
- containing parts starting or ending with `.` (e.g. `refs/branches/.my/branch`).

Also, are we accepting any kind of refs, even if they are ambiguous? (e.g. branch and tag with the same name).

Comment (6543) [reply routers/api/v1/repo/commits.go]: we should think of souch things when **PUSH** ref's but it ref do not exist it simply get a 404 ...

this is only a get func

File: routers/api/v1/repo/commits.go (hunk 2 above)
Comment (6543) [reply routers/api/v1/repo/commits.go]: but then it would be a breaking change!

File: routers/api/v1/repo/commits.go (hunk 3 above)
Comment (lafriks) [reply routers/api/v1/repo/commits.go]: But you could still probably do command injection attack via this probably if anything is accepted and passed to command line

Comment (6543) [reply routers/api/v1/repo/commits.go]: then it already could be done **NOW**

Comment (6543) [reply routers/api/v1/repo/commits.go]: ref: `trash; git execute this!!`

Comment (lafriks) [reply routers/api/v1/repo/commits.go]: not so easy, we probably do some sanity tests for command line arguents but anyway we should allow only sha here not branch names, tags etc

Comment (6543) [reply routers/api/v1/repo/commits.go]: @lafriks `GetSingleCommitBySHA` should be restricted to sha then and `GetSingleCommitByRef` has to check if its a valid rev ...

Comment (zeripath) [reply routers/api/v1/repo/commits.go]: What is a valid git commit reference is really rather difficult to ascertain. If you only want SHA then that's fine however, if you want to allow branches - then... the only answer is to pass the string to `git rev-parse --verify`. This seems mostly safe - remotes are available to be queried however.

Comment (zeripath) [reply routers/api/v1/repo/commits.go]: @6543 

> ref: `trash; git execute this!!`

It doesn't work like that the - the `trash; git execute this!!` is passed to git as a single argument - equivalent to `git rev-parse --verify "trash; git execute this!!"`

**Comment by 6543** [comment]
@zeripath @lafriks @guillep2k so the summary of the discusion is ...

we can pass the string as ref as it is.

and if we restrict `GetSingleCommitBySHA` this will be a breaking change in API :(

File: routers/api/v1/repo/commits.go (hunk 3 above)
Comment (mrsdizzie) [reply routers/api/v1/repo/commits.go]: https://github.com/go-gitea/gitea/blob/master/modules/validation/binding.go has git ref checking if needed

Comment (6543) [reply routers/api/v1/repo/commits.go]: this would need a litle refactor so this func could be used outside macaron

**Comment by techknowlogick** [comment]
This will aid with a Drone feature that is not implemented for Gitea: https://github.com/drone/go-scm/pull/50

**Comment by guillep2k** [comment]
I maintain my opinion that _some_ sanitation should be performed 🙄.  Commits are hex strings from 4 (min) to 40 (max) chars, all lower-case (regexp: `^[0-9a-f]{4,40}$`). As for other references (in the other endpoint), I've listed a couple of cases that should be rejected as git itself rejected them when I tried them from the command line.

**Comment by 6543** [comment]
THIS is ready to review now

**Review by zeripath — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
ping lgtm

**Comment by 6543** [comment]
ready to merge :rocket:

**Comment by lafriks** [comment]
Breaking functionality resolved in #11368

**Pull request merged**